File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # count number of nodes
        temp = head
        noOfNodes = 0
        while temp:
            noOfNodes += 1
            temp = temp.next
        if noOfNodes <= 1:
            return head
        half = math.ceil(noOfNodes / 2)

        # devide the list into two almost equal part
        tail = head
        temp = None
        while half:
            if half == 1:
                temp = tail
            tail = tail.next
            half -= 1
        temp.next = None

        # reverse 2nd list
        prev, curr, nxt = None, tail, tail.next
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        tail = prev


        res = dummy = ListNode()
        flipFlop = 1

        while True:
            if flipFlop % 2 == 1 :
                print('flip ', head.val)
                dummy.next = head
                head = head.next
            else:
                print('flop', tail.val)
                dummy.next = tail
                tail = tail.next
                # print('fgp ', tail.val)
            dummy = dummy.next
            print('ff ',flipFlop)
            if tail == None:
                break
            flipFlop += 1
        
        dummy.next = head

        return res.next
        
        

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverse(ListNode* head, ListNode* tail) {
        ListNode* tailNext = tail->next;

        ListNode* prev = tailNext;
        ListNode* next = head;
        ListNode* curr = head;
        while(curr != tailNext) {
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode *dummy = new ListNode(0, head);
        ListNode *curr = dummy, *tail = dummy;

        while(true) {
            tail = curr;
            for(int i = 0; i < k; i++) {
                tail = tail->next;
                if(tail == nullptr) {
                    return dummy->next;
                }
            }

            ListNode* groupHead = curr->next;
            curr->next = reverse(groupHead, tail);

            curr = groupHead;
        }
        return dummy->next;
    }
};
